### DISCUSSION TITLE
Store data across multiple plot cards?

### DISCUSSION THREAD

**Discussion by FlorianOosterhof** [dbuezas/lovelace-plotly-graph-card#489] [Q&A] [answered]
I have a concept for a dashboard for viewing and comparing heating efficiency over and between long periods of time. For this I need (statistic) history from a long period of time (multiple years) with high precision (daily at least, but preferably even hourly). I wish to have multiple graphs that post-process and slice this data in several ways.

One idea I had was to get the data once, store it in a global way (perhaps a web-page-wide or dashboard-wide JavaScript variable?) that is obtainable by all other cards. This because the data is simply very much, and obtaining it again for each card is costly in terms of time/responsiveness and bandwidth.

Is this possible?

**Comment by dbuezas** [comment, marked as answer]
It is but it will be a hack that may work a bit weird

In the first card write a function that puts the data inside `window.my_data`
Then in the others use empty strings for the entity and load the data from window

**Reply by FlorianOosterhof** [reply]
Thanks for your very quick reply! I will try it out somewhere in the next days and try to make a nice showcase of the result.

~~Can you perhaps help me a bit with the details of how to store the data? I am not very proficient in JS/TS. Would it be as simple as:~~
```
fn: |-
  $ex window.my_data.test_text = "testing"
```

~~or do I need specific JS/TS functions to access this data, similar to the `getFromConfig` function?~~

Note to self: Dont ask questions to which you can test/find the answer quickly.
The above kinda works, but not completely:
```
window.my_test = "test" // works
window.my_data.test_text = "testing" // does not work: my_data does not exist (yet)
window.my_data = {"test_text": "testing"} // works, and obtainable through window.my_data.test_text
```

Thanks again!

**Reply by dbuezas** [reply]
Exactly like that!
You could put this at the beginning of the first card:

```yaml
fn: $ex window.my_vars = vars
```

And this in the following ones:
```yaml
fn: $ex vars = window.my_vars
```

With that you can save/load data with the filters `store_var`. And `load_var` and that would act as a link.

Watch out though, the other cards won't know when to update so you may need to set `refresh_interval: 10s` on the other ones or something like that.
Maybe you could even make the interval variable depending on if the data was found or not (using $ex again)

**Reply by FlorianOosterhof** [reply]
Your construction with storing `vars` globally, will that depend on execution order? Or is this the kind of mechanism where `vars` is stored "by reference" in `window`, so that if `vars` is updated in the first card later on, it is automatically also updated in `window.my_vars` and thus in the other cards `vars`?

Or is it a copy, and thus execution-order dependent? In which case, is there any guarantee on said execution order?

**Reply by dbuezas** [reply]
it's by reference, so it is not important in the first card. It is important that it is very high in the other cards though.

I would just put that at the top on all of them so you don't have to think about it

**Reply by FlorianOosterhof** [reply]
I tried your suggestion, but it seems like it doesnt work as hoped. The second card is done plotting way before the first card is, probably because the first card is still receiving/parsing the data multiple years. I dont know how, but sometimes I manage to get the second card to replot, and then the data appears, so the data sharing works. But the second card should wait until the data is there. Do you know of a way to do that? I couldnt find a way for the JS to sleep, except an `async` one, but apparently this code is synchronous or I did something wrong.

**Reply by dbuezas** [reply]
Yeah, there are no hooks to trigger rerenders from the yaml or $ex. Everything except click handlers results in a single plain object that is passed to plotly and that's the end of it.

The only option you have is to make the second card poll until it has the data:

```yaml
refresh_interval: |
   $ex window.my_vars ? '24h' : '1s'
```
